### PR TITLE
Make it possible to disable bazel symbol stripping

### PR DESCRIPTION
**Pull request by samikama**
Recent changes added build --strip=always to the .tf_configure.bazelrc file which is overriding command-line options for stripping. User has to manually edit the file and remove the flag in order to be able to control strip rules. This PR adds a check that if BAZEL_NO_STRIP is defined in environment, it doesn't add strip=always flag to bazelrc file.

### CODE REVIEW

**Comment by caisq** [comment]
Gentle ping @yifeif

File: configure.py [hunk 1]
```diff
@@ -1523,7 +1523,8 @@ def main():
 
   set_grpc_build_flags()
   set_cc_opt_flags(environ_cp)
-  set_build_strip_flag()
+  if "BAZEL_NO_STRIP" not in environ_cp:
```
Comment (yifeif) [inline configure.py]: Can we do environ_cp.get('BAZEL_NO_STRIP') == '1' instead just to be safe?

Comment (samikama) [reply configure.py]: @yifeif That would imply BAZEL_NO_STRIP=* would keep strip in, which would be counter-intuitive  I guess. How about environ_cp.get('BAZEL_STRIP') == '0'?

Comment (yifeif) [reply configure.py]: Ah but we want the default of stripping bazel flags to stay the same without setting an env var everywhere. What you have sounds good then.

**Comment by samikama** [comment]
@yifeif,

Is this better?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by yifeif — DISMISSED** [review]
Sorry this fell off my radar. Looks good and thanks @samikama!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by perfinion** [comment]
Thanks for this! I originally did the --strip=always fix for an older version but didnt get around to adding a configure flag for it.

Isn't it cleaner to put the check inside set_build_strip_flag()?
also the fix commits should be squashed down i think.

something like:
```
diff --git a/configure.py b/configure.py
index f97bf8a668..884e50d741 100644
--- a/configure.py
+++ b/configure.py
@@ -1399,7 +1399,8 @@ def set_grpc_build_flags():
 
 
 def set_build_strip_flag():
-  write_to_bazelrc('build --strip=always')
+  if environ_cp.get("BAZEL_STRIP") != "0":
+    write_to_bazelrc('build --strip=always')
 
 
 def set_windows_build_flags(environ_cp):
```

**Comment by perfinion** [comment]
I submitted an alternative to this in https://github.com/tensorflow/tensorflow/pull/21373

**Comment by samikama** [comment]
@perfinion,

I was wondering why always strip? Why bazel default is not good? i.e. strip in opt and fastbuild and not strip in debug. What was the rationale behind strip=always?

Thanks?

**Comment by perfinion** [comment]
@samikama its from https://github.com/tensorflow/tensorflow/pull/19599
Originally the linker flags had "-s" in the linkopts in the BUILD files. I removed those cuz things shouldn't be stripped. Martin said they had build issues without it so the compromise was to remove the -s linker flags and add in build --strip=always instead. Sed'ing out the line in one place was much easier to deal with than having patches I'd need to keep resolving conflicts for every release.

I got that PR merged right before the branch so I didn't want to make too big of a change last minute and then forgot to add the configuration option later on.

**Comment by drpngx** [comment]
@perfinion could you pull rebase and push again?

**Comment by perfinion** [comment]
@drpngx PR 21373 has my other version of this. That one is ready to merge so once its in this can be closed.

**Comment by samikama** [comment]
@perfinion,

To be honest, instead of this hack, I would prefer to keep build system flexibility in place. there should be a <name>.stripped option automatically generated by bazel to strip debug symbols. Dependencies could be updated to use stripped binaries instead of this hack.

I would be very much in favor of *not* adding --strip=always to bazelrc

**Comment by perfinion** [comment]
@samikama yeah, I'd rather not have it too. But lets see what @martinwicke says, he asked for it initially.

**Comment by caisq** [comment]
@samikama Sorry for the delay in working on this. Can you resolve the conflict in this PR so we can get it merged? Thanks.

**Comment by samikama** [comment]
@caisq,

Lets wait @martinwicke to answer. It is better if we can have a proper fix instead of hacks and workaround hacks.

**Comment by martinwicke** [comment]
What happens to the build if we disable `strip=always`? will it still build? What is the size of the resulting binaries? Assuming it's all fine (lots of things have changed since the list time we tried this), then I'm fine removing the strip. One way to find out would be to make a PR (maybe this one) which just removes the stripping, and carefully check the sizes of the artifacts (and their build times). That may mean adding an `ls -l` to the build script to make sure we see the size.

**Comment by samikama** [comment]
@martinwicke, 
According to the https://docs.bazel.build/versions/master/user-manual.html#flag--strip, unless it is told to do so or build with dbg flag it should strip.

**Comment by perfinion** [comment]
@martinwicke fwiw, I have disabled stripping on the Gentoo packages since I first packaged it and have not had any reports from users about problems with it.

**Comment by martinwicke** [comment]
Do you know what happens to binary size? We're at the limit of what we can put on pypi. If all is as expected, we should already be stripping, and there's no problem. Let's try it?

**Comment by perfinion** [comment]
@martinwicke @samikama 
They correspond to nothing in .bazelrc, --strip=never, and --strip=always
```
$ ls -l /tmp/dst-*strip/*.whl
-rw-r--r--. 1 jason users 50544207 Aug 14 20:51 /tmp/dst-neverstrip/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
-rw-r--r--. 1 jason users 50544200 Aug 14 20:45 /tmp/dst-nostrip/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
-rw-r--r--. 1 jason users 50544200 Aug 14 20:48 /tmp/dst-strip/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
```
Interestingly, I went back and looked at 1039ff9ee8c8c7ed09f9bb106131a50285866dd4 which was the commit where I removed the `-s` flags. Those were only ever on libtensorflow{,_cc}.so. Not on libtensorflow_framework.so. The wheels only contain the _framework.so. The .so and _cc.so are not distributed on PyPI so shouldn't actually have made a difference anyway.

The *really* heavy handed way to strip is `--linkopt=-Wl,--strip-all`, but even that one only saves ~3.8MB on the entire wheel so also isn't worth the downsides.

**Comment by perfinion** [comment]
@samikama do you want to re-do this PR to remove the stripping? I'll drop the BAZEL_STRIP commit from my other PR then since its pointless.

**Comment by perfinion** [comment]
Heh, if you really want to drop the wheel size, unbundling so far drops >10MB.
```
$ ls -l /tmp/dst-*/*.whl
-rw-r--r--. 1 jason users 50544207 Aug 14 20:51 /tmp/dst-neverstrip/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
-rw-r--r--. 1 jason users 50544200 Aug 14 20:45 /tmp/dst-nostrip/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
-rw-r--r--. 1 jason users 50544200 Aug 14 20:48 /tmp/dst-strip/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
-rw-r--r--. 1 jason users 46485258 Aug 14 21:18 /tmp/dst-stripall/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
-rw-r--r--. 1 jason users 39503309 Aug 14 21:43 /tmp/dst-unbundle/tensorflow-1.10.0-cp36-cp36m-linux_x86_64.whl
```

@martinwicke If you are having issues with PyPI another thing to look into would be building the releases with "-Os" ie optimize for size, I haven't tried but wouldn't be surprised if it saves a fair amount.

**Comment by martinwicke** [comment]
Ok. Can you make this PR (or a new one if you prefer) entirely remove the forced stripping? Then we run the tests and if they're good, we can merge this. If it works, I prefer it over adding another config option. 

cc/ @gunan

**Comment by martinwicke** [comment]
Should we remove `set_build_strip_flag()` entirely?

**Comment by samikama** [comment]
@martinwicke,
Done

**Comment by martinwicke** [comment]
@aselle (or someone you tell me knows) could this tflite profiler failure be related to changes in binary size, or differences in when stripping is performed during the build? Or is it a flake? Time measurements always smell like a flake, but who knows...

**Pull request merged**